Add helpers and checks for new hyperparameter search

samronsin opened the pull request:
Creating the PR to collect feedback, still WIP.
I avoided creating specific classes for each algorithms, while implementing the goals we had initially:
- filter out unmodifiable settings
- catch mistakes early to avoid dreaded JSON parse error from backend
- by having setters for all dimensions
- improve discoverability of hyperparameters
- improve ease of handling hyperparameter search settings (WIP)

Could add helpers for:
- [x] `lr_settings.penalty.set_values({"l1": {"enabled": True}})` (heavy structure)
- [x] `lr_settings.multi_class.set_values("multinomial")` (admissible values check)
- [x] quick getter for dict of active algorithms

[ch53507]

TODO: 
- [x] Doctrings
- [ ] CI to assert e.g. that `save` works as expected ?

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -246,12 +246,22 @@ def get_algorithm_settings(self, algorithm_name):
+        raw_algorithm_settings = self.mltask_settings["modeling"][algorithm_name.lower()]
+        raw_hyperparameter_search_settings = self.mltask_settings["modeling"]["gridSearchParams"]
+
+        if algorithm_name in {"random_forest_classification", "random_forest_regression", "extra_trees"}:
+            return RandomForestSettings(raw_algorithm_settings, raw_hyperparameter_search_settings)
+        elif algorithm_name == "xgboost":
+            return XGBoostSettings(raw_algorithm_settings, raw_hyperparameter_search_settings)
+        elif algorithm_name == "logistic_regression":
+            return LogitSettings(raw_algorithm_settings, raw_hyperparameter_search_settings)
+        else:
+            return AlgorithmSettings(raw_algorithm_settings, raw_hyperparameter_search_settings)
```
Yes, we should probably use the already existing `algorithm_remap`.

samronsin replied:
I tried a new design, feedback very welcome !

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,13 +318,296 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    __str__ = __repr__
+
+    def _get_clean_numerical_hyperparam_dict(self, hyperparameter_name):
```
How about `_get_printable_numerical_hyperparam_dict ` ?

instanceofme replied:
It's kind of in between:
- on one hand, you could say that `printable` is the function, and `active` is an implementation detail
- on the other hand, you could say that `active` is what it does, and it is agnostic of what you use it for…

So I guess it depends on whether we think this behavior may have other uses than print. I tend to think yes, but that's a loosely held opinion.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,13 +318,296 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            return self.raw_settings[hyperparameter_name]["randomMode"]
+
+    def _check_hyperparameter_name(self, hyperparameter_name):
+        assert isinstance(hyperparameter_name, str), "Invalid type for hyperparameter name: expecting a string"
```
Good catch, thanks !

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,13 +318,296 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        elif isinstance(self.raw_settings[hyperparameter_name], int) or isinstance(self.raw_settings[hyperparameter_name], float):
+            assert isinstance(value, int) or isinstance(value, float), "Invalid input type for hyperparameter \"{}\": expected a number".format(hyperparameter_name)
+        else:
+            raise ValueError("Hyperparameter \"{}\" is not single valued".format(hyperparameter_name))
```
I assume you mean this for the first `if` here ?

instanceofme replied:
Initially yes, but actually maybe both, I don't know how `format` deals this that?

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -364,6 +364,14 @@ def __repr__(self):
     def set_values(self, values=None):
         self._algo_settings._set_categorical_values(self.name, values=values)
 
+    def enable_values(self, values_list):
+        for value in values_list:
+            self.set_values({value: {"enabled": True}})
+
+    def disable_values(self, values_list):
+        for value in values_list:
+            self.set_values({value: {"enabled": False}})
```
Why not just `values` instead of `values_list`?

samronsin replied:
It's just an old habit of mine of adding type to the name of variables, but I guess here we could accept any iterable and not only lists.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +331,554 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+    def set_values(self, values=None):
+        self._algo_settings._set_categorical_values(self.name, values=values)
+
+    def enable_values(self, values):
```
How about an additional argument `only=False` that, when set to `True`, would disable all others?

Otherwise if you want values `A` and `B` and no others, you need to know all the possible values (not future friendly as adding new values may ~break old code).

Would also be added to `disable_values` and work the opposite.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +331,554 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self._algo_settings._set_single_valued_hyperparameter(self.name, value)
+
+
+def is_str_or_unicode(val):
+    import sys
+    if sys.version_info < (3, 0):
+        return isinstance(val, basestring)
+    else:
+        return isinstance(val, str)
```
This should probably be in some utils module

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +331,554 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self.accepted_values = accepted_values
+
+    def __repr__(self):
+        return self.__class__.__name__ + "\nhyperparameter: \"{}\"".format(self.name) + "\nvalue: " + json.dumps(self._algo_settings[self.name], indent=4)
```
```
SingleValuedHyperparameterSettings
hyperparameter: "multi_class"
value: "ovr"
```

Maybe we could instead have a single line?

```
SingleValuedHyperparameterSettings(hyperparameter: "multi_class", value: "ovr")
```

And for multi-line settings, go from
```
CategoricalHyperparameterSettings
hyperparameter: "penalty"
settings: {
    "values": {
        "l1": {
            "enabled": true
        },
        "l2": {
            "enabled": true
        }
    }
}
```

To:

```
CategoricalHyperparameterSettings(hyperparameter: "penalty", settings: {
    "values": {
        "l1": {
            "enabled": true
        },
        "l2": {
            "enabled": true
        }
    }
})
```

instanceofme commented:
Things to do once finalized:

- [ ] Write the missing docstrings
- [ ] Change base branch to `master`

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        assert isinstance(self[hyperparameter_name], dict), "Hyperparameter \"{}\" cannot be searched".format(hyperparameter_name)
+
+    def _check_hyperparameter_numerical(self, hyperparameter_name):
+        assert "range" in self[hyperparameter_name], "Hyperparameter \"{}\" is not numerical or cannot be searched".format(hyperparameter_name)
```
```suggestion
        assert "range" in self[hyperparameter_name], "Hyperparameter \"{}\" is not numerical".format(hyperparameter_name)
```

Always called after `_check_hyperparameter_searchable` so we now that it is searchable

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+    def _check_hyperparameter_numerical(self, hyperparameter_name):
+        assert "range" in self[hyperparameter_name], "Hyperparameter \"{}\" is not numerical or cannot be searched".format(hyperparameter_name)
+
+    def _check_range_bound(self, hyperparameter_name, range_bound):
```
Nitpick: Only called in `_set_numerical_range`. Why not explicitly `_check_hyperparameter_numerical` before instead of rechecking if `"range"` is there with a different assertion error message ?

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            assert isinstance(values, dict), "Invalid values input type for hyperparameter " \
+                                             "\"{}\": ".format(hyperparameter_name) + \
+                                             "must be a dictionary"
+            admissible_values = list(self[hyperparameter_name]["values"].keys())
```
Where are these admissible_values first defined ? In the backend ?

instanceofme replied:
Yes

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self.base_score = SingleValuedHyperparameterSettings("base_score", self)
+        self.impute_missing = SingleValuedHyperparameterSettings("impute_missing", self, [True, False])
+        self.missing = SingleValuedHyperparameterSettings("missing", self)
+        self.cpu_tree_method = SingleValuedHyperparameterSettings("cpu_tree_method", self)
+        self.gpu_tree_method = SingleValuedHyperparameterSettings("gpu_tree_method", self)
```
For these, we have no accepted values, because it is a space delimited string, right ?

samronsin replied:
No I just forgot about them !

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+    def __init__(self, name, algo_settings, accepted_values=None):
+        super(SingleValuedHyperparameterSettings, self).__init__(name, algo_settings)
+        self.accepted_values = accepted_values
+
```
Wouldn't it be relevant to add an accepted_type for the `SingleValuedHyperparameterSettings`, since they can be numerical or not (string etc)

samronsin replied:
Yes totally -- already explicitly checked for `bool` but adding the type would be better.

mhham reviewed (changes requested):
A few remarks:
* A bit of naming improvement (mainly `NamedAlgorithm`)
* For the organization of `AlgorithmSettings` and `HyperparameterSettings` , I am a bit puzzled by the fact that we put all the logic in `AlgorithmSettings` and not in the relevant `HyperparameterSettings` subclasses, and thus have to pass the whole `AlgorithmSettings` object in the `HyperparameterSettings` constructor
* For `SingleValuedHyperparameterSettings` a question about adding an accepted type
* Some small code simplifications
* A question about hyperparams settings that are not dicts
* Some `__repr__` function are multi-line, is this good for logs etc ?
...

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            res += self._key_repr("randomized")
+            if self._raw_settings.get("randomized", False):
+                res += self._key_repr("seed")
+        else:
```
Yes, to be added in comment I guess.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            res += self._key_repr("seed")
+
+        res += "Parallelism Settings:\n"
+        res += self._key_repr("nJobs")
```
Yes, the naming is a bit misleading -- we could add inline comments in the repr to help the user disambiguate.

instanceofme replied:
Not sure we want to go as far as to document this in the repr.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self.step_size = NumericalHyperparameterSettings("step_size", self)
+
+
+class NamedAlgorithm:
```
Among proposed names, a marked preference for the simpler `AlgorithmMeta`

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -292,6 +299,23 @@ def get_all_possible_algorithm_names(self):
+        :returns: the list of enabled algorithm names as a list of strings
+        :rtype: list of string
+        """
+        return [key for key in self.__class__.algorithm_remap.keys() if self.get_algorithm_settings(key)["enabled"]]
```
> Can't we simply access class attributes with: self.class_attribute ?
It works but is considered bad practice, as theoretically an instance could have a different value overriding that.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    def __init__(self, name, algo_settings):
+        self.name = name
+        self._algo_settings = algo_settings
```
Not sure about that second proposal, seems more complex.

instanceofme replied:
Not sure I get it, `AlgorithmSettings` _is_ the hyperparameter settings? It allows to keep the reference (not copy of the whole thing, just a reference) on the rest so that you can filter out irrelevant (inactive) parts in the repr.

Extending `dict` is for backward compatibility, not sure we want to do it more than necessary unless we have a good reason to do so.

mhham replied:
I agree that the current code only keeps the reference (not the copy) of `AlgorithmSettings`. 

It's just a matter of design, since I was intuitively thinking about `HyperparameterSettings` as being part of the `AlgorithmSettings`, while the current state is that all the info is in `AlgorithmSettings`, and `HyperparameterSettings` only provide the methods to correctly print / modify the relevant parts in the `AlgorithmSettings` object.

In the end, the API usage would remain the same,

samronsin replied:
@mhham I think the original reason was that I needed the `strategy` to define the actual `mode` (either `gridMode` or `randomMode`) in the `HyperparameterSettings`, so I just kept the pointer to the larger object with the name of the hyperparameter field.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    __str__ = __repr__
+
+    def set_explicit_values(self, values, mode_as_explicit=True):
```
No strong opinion about the 2nd parameter.

A point about the name though (and same for `set_mode_as_*` and `_set_hyperparameter_mode_as_*`): in the rest of the code, `_as_*` is used for casting / type. Not sure we want to divert from this, maybe we can find an alternative name? `set_explicit_mode`, `set_mode_to_explicit`?

samronsin replied:
Regarding `*_as_*`, I didn't pay attention but I agree on enforcing a consistency. Since `set_mode_to_explicit` is already used as a method name, I suggested `mode_becomes_explicit` and `mode_becomes_range`.

instanceofme replied:
Sound a bit weird, doesn't it? I don't have a problem with using the same name as a method if it does the same thing / calls the method (which it seems to do internally).

On the other hand, don't we want to use the same rationale as the validation mode and grid search method, and simplify to a single method for explicit + values, one for range + limits, etc? (And potentially keep gettable+settable individual properties on the side if we deem it useful).

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        if limit_max is not None:
+            assert all(val <= limit_max for val in values), "Value(s) above hyperparameter \"{}\" limit {}".format(hyperparameter_name, limit_max)
+        if len(set(values)) < len(values):
+                warnings.warn("Detected duplicates in provided values: " + str(sorted(values)))
```
I guess to easily spot the duplicated value?

mhham replied:
In that case, would it be more readable to just print the duplicates ?

instanceofme replied:
Maybe overkill :)

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        return clean_hyperparam
+
+    def _get_all_hyperparameter_names(self):
+        return [key for key in self.keys() if key not in self.excluded_keys]
```
I don't think `list` vs `set` matters at all for the number of hyperparameters at hand (10-ish at most).

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        assert values is not None and isinstance(values, list) and len(values) > 0, error_message
+        for val in values:
+            assert isinstance(val, int) or isinstance(val, float), error_message
+        limit_min = self[hyperparameter_name]["limit"].get("min")
```
All `NumericalHyperparameterSettings` have a `limit` key, otherwise the underlying dict is corrupted.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -246,12 +247,18 @@ def get_algorithm_settings(self, algorithm_name):
+        raw_hyperparameter_search_settings = self.mltask_settings["modeling"]["gridSearchParams"]
 
-        return self.mltask_settings["modeling"][algorithm_name.lower()]
+        return algorithm_settings_class(raw_algorithm_settings, raw_hyperparameter_search_settings)
```
Additional suggestion after team discussion: rename `AlgorithmSettings` to `AlgorithmParams` (and same with `HyperparameterSearchSettings`) to make it clearer:

- `~Settings` is something you need to `save()` individually
- `~Params` or others are not, you `save()` the parent

samronsin replied:
An issue with this renaming is that `get_algorithm_settings` will return an `AlgorithmParams`...

instanceofme replied:
True that 🤔

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        super(AlgorithmSettings, self).__init__(raw_settings)
+        self._hyperparameter_search_params = hyperparameter_search_params
+
+    def __repr__(self):
+        clean_settings = dict()
+        for key in self.keys():
+            if isinstance(self[key], dict):
+                # Searchable hyperparameter
+                if "range" in self[key]:
+                    # Numerical hyperparameter
+                    clean_hyperparam = self._get_active_numerical_hyperparam_dict(key)
+                else:
+                    # Categorical hyperparameter
+                    clean_hyperparam = self[key]
+                clean_settings[key] = clean_hyperparam
+            else:
+                clean_settings[key] = self[key]
+        return self.__class__.__name__ + "(values={})".format(json.dumps(clean_settings, indent=4))
+
+    __str__ = __repr__
```
After discussion, in HTML we can only display the relevant metrics + add a full dict dump in a `<details>` tag:

```html
<details>
<summary>Raw parameters</summary>
<pre>{ ...dump... }</pre>
</details>
```

Looks like this:

<details>
<summary>Raw parameters</summary>
<pre>{ ...dump... }</pre>
</details>

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+            assert is_basestring(value)
+            self.set_values({value: {"enabled": True}})
+        if disable_others:
+            other_values = [key for key in self._algo_settings[self.name]["values"].keys() if key not in values]
+            for value in other_values:
+                self.set_values({value: {"enabled": False}})
```
Thanks for spotting this !

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
         self.mltask_settings["modeling"]["metrics"]["customEvaluationMetricNeedsProba"] = custom_metric_use_probas
 
+    def get_hyperparameter_search_settings(self):
+        return HyperparameterSearchSettings(self.mltask_settings["modeling"]["gridSearchParams"])
```
Indeed, no copy as `HyperparameterSearchSettings` does not use `dict` constructor.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        assert strategy in {"GRID", "RANDOM", "BAYESIAN"}
+        self._raw_settings["strategy"] = strategy
+
+    def set_validation_mode(self, mode):
+        assert mode in {"SHUFFLE", "KFOLD", "TIME_SERIES_SINGLE_SPLIT", "TIME_SERIES_KFOLD", "CUSTOM"}
+        self._raw_settings["mode"] = mode
```
Agreed and modified accordingly.

instanceofme replied:
Not pushed yet right?

samronsin replied:
Yes indeed -- doing it now.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    def __repr__(self):
+        clean_dict = self._algo_settings._get_active_numerical_hyperparam_dict(self.name)
+        return self.__class__.__name__ + "(hyperparameter=\"{}\", settings={})".format(self.name, json.dumps(clean_dict, indent=4))
```
It's debatable but I don't think it's an issue to make the logs _more_ readable, and since they often end up being searched with tools such as less or grep anyways, having more lines should not be a problem.
The idea here is to provide relevant and readable information to the user at a glance, especially when they are designing experiments in notebooks.

instanceofme replied:
Hmm, specifically, if we make grep-ability a priority over readability, that would be an argument for remaining on a single line, as `cat my.log | grep some-expr` will by default only display that line and not display the other lines form that object. Whereas if the whole object is on a single line, it will get displayed. But a single line JSOn is harder to read…

instanceofme replied:
In this case I'd tend to favor greppability / one-line in string repr (precisely because log processing tools work better with one line), but readability / multi-line in HTML for notebooks (including details `<pre>`).

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -393,40 +392,81 @@ def __repr__(self):
+            else:
+                self._raw_settings["seed"] = seed
+
+    def set_strategy_to_grid_search(self, shuffle=None, seed=None):
```
That sounds a bit weird, although I'm not sure I have better. `use_grid_search`? or `set_grid_search`?

instanceofme replied:
`shuffle` and `seed` should default to actual default (if applicable, maybe null `seed` is valid?), makes it simpler & self-documenting.
[Same with all similar methods, including those for validation mode]

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -393,40 +392,81 @@ def __repr__(self):
-        if self._raw_settings["mode"] not in {"KFOLD", "TIME_SERIES_KFOLD"}:
-            warnings.warn("\"nFolds\" parameter will be ignored in {} mode".format(self._raw_settings["mode"]))
-        self._raw_settings["nFolds"] = n_folds
+    def set_validation_mode_to_kfold(self, n_folds=None, stratified=None):
```
`set_validation_kfold`? `set_kfold_validation`?

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -468,26 +486,44 @@ def set_validation_mode_to_custom(self, code=None):
+        if n_containers is not None:
+            assert isinstance(n_containers, int)
+            self._raw_settings["nContainers"] = n_containers
+        self._raw_settings["distributed"] = distributed
```
We may also want at least a reader property for distributed?

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -620,6 +621,8 @@ def __repr__(self):
         if self.accepted_values is not None:
             assert value in self.accepted_values, "Invalid value for hyperparameter {}. Must be in {}".format(self.name, json.dumps(self.accepted_values))
+        if self.accepted_types is not None:
+            assert any(isinstance(value, type) for type in self.accepted_types), "Invalid type for hyperparameter {}. Type must be one of {}".format(self.name, self.accepted_types)
```
```suggestion
            assert any(isinstance(value, type) for type in self.accepted_types), "Invalid type for hyperparameter {}. Type must be one of: {}".format(self.name, self.accepted_types)
```

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        super(AlgorithmSettings, self).__init__(raw_settings)
+        self._hyperparameter_search_params = hyperparameter_search_params
+
+    def _repr_html_(self):
```
This seems risky to iterate over the underlying dict, while we always manipulate fields otherwise.
Wouldn't it be cleaner to have a `register_hyperparameters` called at the end of the init of each algo  (that would also serve for `_get_all_hyperparameter_names`).

This would allow to leverage methods like `pretty_repr` directly written from `NumericalHyperparameterSettings` and `CategoricalHyperparameterSettings` rather than manually building them here.

samronsin replied:
Thanks for this comment, the iteration is quite clumsy indeed.
How about 
```python
class AlgorithmSettings(dict):

    excluded_keys = {"enabled"}

    def __init__(self, raw_settings, hyperparameter_search_params):
        super(AlgorithmSettings, self).__init__(raw_settings)
        self._hyperparameter_search_params = hyperparameter_search_params
        self._hyperparameters_registry = dict()

    def register_numerical_hyperparameter(self, name):
        self._hyperparameters_registry[name] = NumericalHyperparameterSettings(name, self)
        return self._hyperparameters_registry[name]

    def register_categorical_hyperparameter(self, name):
        self._hyperparameters_registry[name] = CategoricalHyperparameterSettings(name, self)
        return self._hyperparameters_registry[name]

    def register_single_valued_categorical_hyperparameter(self, name, accepted_values=None):
        self._hyperparameters_registry[name] = SingleValuedHyperparameterSettings(name, self, accepted_values=accepted_values)
        return self._hyperparameters_registry[name

    def register_single_valued_hyperparameter(self, name, accepted_types=None):
        self._hyperparameters_registry[name] = SingleValuedHyperparameterSettings(name, self, accepted_types=accepted_types)
        return self._hyperparameters_registry[name]

    def _repr_html_(self):
        res = "<pre>" + self.__class__.__name__ + "\n"
        for name, hyperparam_settings in self._hyperparameters_registry.items():
            res += hyperparam_settings._pretty_repr() + "\n"
        return res + "<details><pre>{}</pre></details>".format(self.__repr__())

    def __repr__(self):
        return self.__class__.__name__ + "(values={})".format(json.dumps(super(AlgorithmSettings, self).copy(), indent=4))

    __str__ = __repr__

    def _get_all_hyperparameter_names(self):
        return list(self._hyperparameters_registry.keys())
    ...
```
and then each algo settings would become
```python
class LogitSettings(AlgorithmSettings):

    def __init__(self, raw_settings, hyperparameter_search_params):
        super(LogitSettings, self).__init__(raw_settings, hyperparameter_search_params)
        self.C = self.register_numerical_hyperparameter("C")
        self.penalty = self.register_categorical_hyperparameter("penalty")
        self.multi_class = self.register_single_valued_categorical_hyperparameter("multi_class", accepted_values=["multinomial", "ovr"])
        self.n_jobs = self.register_single_valued_hyperparameter("n_jobs", accepted_types=[int])
```

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+
+    __str__ = __repr__
+
+    def set_mode_to_explicit(self, values=None):
```
[Personal opinion] it feels weird, when I am already in "EXPLICIT" and want to change values, to have to use a method called `set_mode_to_explicit`

samronsin replied:
In the same vein as the new method names for `HyperparameterSearchSettings` (`set_grid_search`...), we could name this one `set_explicit_mode` etc. I guess this would sound less weird ?

instanceofme replied:
`set_explicit_values(self, values)`?

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+    def __init__(self, raw_settings, hyperparameter_search_params):
+        super(RandomForestSettings, self).__init__(raw_settings, hyperparameter_search_params)
+        self.n_estimators = NumericalHyperparameterSettings("n_estimators", self)
+        self.min_samples_leaf = NumericalHyperparameterSettings("n_estimators", self)
```
```suggestion
        self.min_samples_leaf = NumericalHyperparameterSettings("min_samples_leaf", self)
```

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self.colsample_bytree = NumericalHyperparameterSettings("colsample_bytree", self)
+        self.colsample_bylevel = NumericalHyperparameterSettings("colsample_bylevel", self)
+        self.alpha = NumericalHyperparameterSettings("alpha", self)
+        self.lambda_ = NumericalHyperparameterSettings("lambda", self)
```
why the `_` at the end of `lambda_` ?

instanceofme replied:
`lambda` is also a python keyword

samronsin replied:
Actually getting a `SyntaxError` with the reserved keyword (Python 3.6.8).

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+    def __init__(self, raw_settings, hyperparameter_search_params):
+        super(KNNSettings, self).__init__(raw_settings, hyperparameter_search_params)
+        self.k = NumericalHyperparameterSettings("k", self)
+        self.algorithm = CategoricalHyperparameterSettings("algorithm", self)
```
this is not a Categorical, but a SingleValue hp

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+class SVMSettings(AlgorithmSettings):
+
+    def __init__(self, raw_settings, hyperparameter_search_params):
+        super(SVMSettings, self).__init__(raw_settings, hyperparameter_search_params)
```
missing the `C` param

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
         self.mltask_settings["modeling"]["metrics"]["customEvaluationMetricGIB"] = custom_metric_greater_is_better
         self.mltask_settings["modeling"]["metrics"]["customEvaluationMetricNeedsProba"] = custom_metric_use_probas
 
+    def get_hyperparameter_search_settings(self):
```
Let's implement that on the backend (enforce some basic checks on save) to avoid tight coupling.

instanceofme replied:
Including some of the things that we currently check in this PR, like min < max for ranges…

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +334,757 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        if self.accepted_values is not None:
+            assert value in self.accepted_values, "Invalid value for hyperparameter {}. Must be in {}".format(self.name, json.dumps(self.accepted_values))
+        if self.accepted_types is not None:
+            assert any(isinstance(value, type) for type in self.accepted_types), "Invalid type for hyperparameter {}. Type must be one of {}".format(self.name, self.accepted_types)
```
`type` is not a reserved keyword, only a built-in _function_ so using it as a variable is perfectly safe. Switching to `accepted_type` nevertheless as it's just as expressive and won't trigger the code reader.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -245,13 +246,23 @@ def get_algorithm_settings(self, algorithm_name):
+            algorithm_params = algorithm_settings_class(algorithm_settings, raw_hyperparameter_search_params)
+            # Subsequent calls get the same object
+            self.mltask_settings["modeling"][algorithm_name.lower()] = algorithm_params
+        return algorithm_settings
```
Indeed...

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -592,11 +585,121 @@ def _get_active_settings_dict(self):
+                self._algo_settings[self.name]["range"]["max"] = max
+            if nb_values is not None:
+                self._algo_settings[self.name]["range"]["nbValues"] = nb_values
+        if set_mode_to_range:
+            self._set_mode_to_range()
```
Ditto

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -610,7 +713,22 @@ def _pretty_repr(self):
     def set_values(self, values=None):
-        self._algo_settings._set_categorical_values(self.name, values=values)
+        if values is None:
+            warnings.warn("Categorical hyperparameter \"{}\" not modified".format(self.name))
```
I'm not sure I get why we'd want to support `set_values(None)` that doesn't do anything? Isn't it simpler to just force values to be passed?

instanceofme reviewed (commented):
I understood things differently from our last discussion. I'd suggest to align everything to

- properties are low-level getter/setter without other side effects
- `set_some_mode(param1=default, param2=default)` both sets relevant mode (not optional) and parameters for that mode
- if one wants to set things independently, they'd use properties

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -592,11 +585,121 @@ def _get_active_settings_dict(self):
+            return self._algo_settings[self.name]["gridMode"]
+        else:
+            # RANDOM and BAYESIAN search strategies
+            return self._algo_settings[self.name]["randomMode"]
```
It's not really "search" mode though...

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -592,32 +757,45 @@ def disable_categories(self, categories, enable_others=False):
+        self._algo_settings[self.name] = value
+
+
+class SingleValuedCategoricalHyperparameterSettings(HyperparameterSettings):
```
The naming suggest that this inherits from `SingleValuedHyperparameterSettings`, can we find a less ambiguous name for `SingleValuedHyperparameterSettings` or maybe factor common behavior in a base class ?

samronsin replied:
There really is not much in common to factor...
We could name them `SingleValueHyperparameterSettings` and `SingleCategoryHyperparameterSettings` ?

mhham replied:
Looks better !

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -627,46 +805,38 @@ class AlgorithmSettings(dict):
+        return res + "<details><pre>{}</pre></details>".format(self.__repr__())
 
     def __repr__(self):
         return self.__class__.__name__ + "(values={})".format(json.dumps(super(AlgorithmSettings, self).copy(), indent=4))
```
Shouldn't we have a single line `__repr__` now that we have the `_repr_html_` ?

samronsin replied:
Indeed, thanks !

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -292,6 +303,21 @@ def get_all_possible_algorithm_names(self):
         """
         return self.__class__.algorithm_remap.keys()
 
+    def get_enabled_algorithm_names(self):
+        """
+        :returns: the list of enabled algorithm names as a list of strings
+        :rtype: list of string
+        """
+        algos = self.__class__.algorithm_remap
+        return [algorithm_name for algorithm_name in algos.keys() if self.mltask_settings["modeling"][algos[algorithm_name].algorithm_name.lower()]["enabled"]]
+
+    def get_enabled_algorithm_settings(self):
```
Beware that you may have, for a single task:
- Several custom models (instantiating code inline)
- Several plugin models

samronsin replied:
I ended up adding an entry for each of these, with value the raw list of dicts, so the user could possibly enable/disable them. Not sure it's worth making wrapper classes for them (maybe for each list, to enable/disable all elements though ?): the only upside I can think of is to discourage the user from modifying the code via the API which would already be quite cumbersome. What do you think ?

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -257,12 +259,12 @@ def get_algorithm_settings(self, algorithm_name):
-        raw_hyperparameter_search_params = self.mltask_settings["modeling"]["gridSearchParams"]
         if not isinstance(algorithm_settings, AlgorithmSettings):
+            raw_hyperparameter_search_params = self.mltask_settings["modeling"]["gridSearchParams"]
             algorithm_params = algorithm_settings_class(algorithm_settings, raw_hyperparameter_search_params)
```
OK indeed since it's an instance of `AlgorithmSettings`, let's name this `algorithm_settings`.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -534,41 +551,189 @@ def __init__(self, name, algo_settings):
+
+    def __init__(self, numerical_hyperparameter_settings):
+        self._numerical_hyperparameter_settings = numerical_hyperparameter_settings
+        self._algo_settings = numerical_hyperparameter_settings._algo_settings
```
True dat, I forgot to clean this up !

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -534,41 +551,189 @@ def __init__(self, name, algo_settings):
+                assert isinstance(nb_values, int) and nb_values >= 2, "Range number of values for hyperparameter \"{}\" must be an integer and >= 2".format(self.name)
+
+            if min is not None:
+                self._algo_settings[self.name]["range"]["min"] = min
```
Actually, I don't understand what you are suggesting: 
- all setters of `Range` end up calling this method, so this would mean swapping code snippets between `NumericalHyperparameterSettings._set_range` and the setters of `Range` and I'd rather keep the complexity in one entrypoint (`_set_range`)
- we don't want to change the type of `self._algo_settings[self.name]["range"]` (a `dict`).

mhham replied:
Actually forget it. I thought the `Range` class would contain all the methods to modify the range (a bit like the various `HyperparameterSettings` classes contain the methods to modify the hyperparams settings).

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -251,20 +251,22 @@ def get_algorithm_settings(self, algorithm_name):
         :return: An AlgortihmSettings (extended dict) for an algorithm.
         :rtype: AlgorithmSettings
```
So return type may also be `dict` (also: typo)

```suggestion
        :return: An AlgorithmSettings (extended dict) for a single built-in algorithm,
                 or a plain dict for the settings of custom inline-code (CUSTOM_*) or plugin (PLUGIN_*) algorithms.
        :rtype: AlgorithmSettings | dict
```

Wondering whether we _also_ want to specify those in the docstring, or if it just makes it more complicated.

samronsin replied:
I think that `"CUSTOM_*"` and `"PLUGIN_*"` are not really 1st-class arguments of this method, so I'd rather keep the documentation as clear as possible for the vast majority of usages and be slightly non-exhaustive in the docstring explanation.

samronsin replied:
Marking as resolved as custom and plugin algorithms will be handled separately.

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -245,13 +248,25 @@ def get_algorithm_settings(self, algorithm_name):
+        elif algorithm_name in self.__class__.algorithm_remap:
+            algorithm_meta = self.__class__.algorithm_remap[algorithm_name]
+            algorithm_name = algorithm_meta.algorithm_name
+            algorithm_settings_class = algorithm_meta.algorithm_settings_class
```
Raising the error for unknown algorithm names is still relevant

```suggestion
            algorithm_settings_class = algorithm_meta.algorithm_settings_class
        else:
            raise ValueError("Unknown algorithm: {}".format(algorithm_name))	
```

instanceofme reviewed (commented):
Starting to look really nice, some nitpicks & minor stuff

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,951 @@ def save(self):
+        return res + "<details><pre>{}</pre></details>".format(self.__repr__())
+
+    def __repr__(self):
+        return self.__class__.__name__ + "(values={})".format(super(PredictionAlgorithmSettings, self).copy())
```
Why do we need to call the super method here, as it's not overriden in the class ?

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -230,28 +233,7 @@ def use_feature(self, feature_name):
-            algorithm_name = self.__class__.algorithm_remap[algorithm_name]
-
-        return self.mltask_settings["modeling"][algorithm_name.lower()]
+        raise NotImplementedError()
```
Shouldn't we deprecate this method, rather than breaking it ? Note that it currently also breaks `set_algorithm_enabled`

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -234,24 +237,37 @@ def get_algorithm_settings(self, algorithm_name):
+        :rtype: AlgorithmSettings | dict
+        """
+        if algorithm_name in ["CUSTOM_MLLIB", "CUSTOM_PYTHON", "PLUGIN_PYTHON"]:
+            return self.mltask_settings["modeling"][algorithm_name.lower()]
```
After discussion: not support custom & plugins for now, add those in a separate PR.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -698,28 +773,45 @@ def set_values(self, values=None):
+                              for category in accepted_categories
+                              if category not in categories})
 
     def disable_categories(self, categories, enable_others=False):
```
After discussion, align single- and multi-valued HP's method + consistency with algo list + syntactic sugar (through `__setattr__` on the parent):
- Single-valued HP: `single.get_all_possible_values()` / `get_value()` / `set_value(v)` / `single = v`
- Multi-valued HP:  `multi.get_all_possible_values()` / `get_values()` / `set_values([v1, v2])` / `multi = [v1, v2]`  
  Setter disables other possible values.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,951 @@ def save(self):
+            res += self._key_repr("nFolds")
+
+        res += self._key_repr("stratified")
+        # TODO: pointers to ssdSeed and time variable ?
```
More of an open question. 
They are not direct entries of the underlying `dict` (`_raw_settings`), but accessible in the larger `MLTaskSettings`. Displaying the time variable and the seed used for train/test split and also for validation splits will probably add value to the user (at the cost of keeping a pointer to the parent `MLTaskSettings` in the current `HyperparameterSearchSettings`), but setting them may not be appropriate here. We could do the same as with `strategy` in `PredictionAlgorithmSettings` and have a read-only property with pointer to the relevant setters.

mhham replied:
IMO, keeping a pointer to the `MLTaskSettings` only to expand the `__repr__` is a bit overkill, especially that these variables are directly accessible through the `MLTaskSettings`.
Do you have a specific use case where the user will only print the Algo settings ?

samronsin replied:
I created a [CH card](https://app.clubhouse.io/dataiku/story/58357/make-the-seed-of-the-hyperparameter-search-independent-from-the-seed-of-the-train-test-split) to address the `seed` issue, which is currently a bit dysfunctional IMO.

Regarding the time ordering variable, I agree that there is not much room for handling it from the `HyperparameterSearchSettings` instead of the dedicated `PredictionSplitParamsHandler`.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -420,17 +435,27 @@ def validation_mode(self):
     def set_kfold_validation(self, n_folds=5, stratified=True):
+        """
+        Sets the validation mode to k-fold cross-validation (either "KFOLD" or "TIME_SERIES_KFOLD" if time-based ordering
+        is enabled).
```
There are two helpers in `PredictionSplitParamsHandler`: `set_time_ordering` and `unset_time_ordering`... We could add a read-only property in the `HyperparameterSearchSettings`.

samronsin replied:
We could also add a simple `time` property in `PredictionSplitParamsHandler`, so all time-based ordering matters would be located in the split params object.

samronsin replied:
Or simply (in `PredictionSplitParamsHandler`):
```python
    def get_time_ordering_variable(self):
        """
        :return: the name of the variable
        :rtype: str
        """
        if self.mltask_settings.get('time', None) and self.mltask_settings['time']['enabled']:
            return self.mltask_settings['time']['timeVariable']
        else:
            warnings.warn("Time-based ordering is disabled in the current MLTask")

    def is_time_ordering_ascending(self):
        """
        :return: True if the ordering is set to be ascending with respect to the time-ordering variable
        :rtype: bool
        """
        if self.mltask_settings.get('time', None) and self.mltask_settings['time']['enabled']:
            return self.mltask_settings['time']['ascending']
        else:
            warnings.warn("Time-based ordering is disabled in the current MLTask")

```

samronsin replied:
As discussed, to be done in a separate PR (created [CH58931](https://app.clubhouse.io/dataiku/story/58931/add-getter-s-for-time-based-ordering-in-python-client-of-public-api)).

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -946,7 +984,7 @@ def _repr_html_(self):
 
     def __repr__(self):
-        return self.__class__.__name__ + "(values={})".format(super(PredictionAlgorithmSettings, self).copy())
+        return self.__class__.__name__ + "(values={})".format(self.copy())
```
Why a copy ?

instanceofme replied:
My guess: because `copy` produces a dict with its regular string repr, while `self` would look to convert `self` to a string, recursively calling `__repr__`, causing infinite recursion.

mhham replied:
Then maybe:
```suggestion
        return self.__class__.__name__ + "(values={})".format(dict.__repr__(self))
```
or
```suggestion
        return self.__class__.__name__ + "(values={})".format(super(PredictionAlgorithmSettings, self).__repr__())
```

mhham replied:
> My guess: because `copy` produces a dict with its regular string repr, while `self` would look to convert `self` to a string, recursively calling `__repr__`, causing infinite recursion.

Also: when calling `"{}".format(something)` this calls `something.__str__()`. In this case it is the same as `something.__repr__()` because `__str__ = __repr__` when it is not explicitly implemented

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -1449,7 +1486,8 @@ def get_algorithm_settings(self, algorithm_name):
         Please refer to the documentation for details on available algorithms.
 
-        :param str algorithm_name: Name (in capitals) of the algorithm.
+        :param: algorithm_name: Name (in capitals) of the algorithm.
```
```suggestion
        :param: algorithm_name: Name of the algorithm (uppercase).
```

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -1297,14 +1336,12 @@ def get_algorithm_settings(self, algorithm_name):
-        :rtype: PredictionAlgorithmSettings | dict
+        :param algorithm_name: Name (in capitals) of the algorithm.
+        :type algorithm_name: str
+        :return: A PredictionAlgorithmSettings (extended dict) for a single built-in prediction algorithm
```
```suggestion
        :return: A PredictionAlgorithmSettings (extended dict) for one of the built-in prediction algorithms
```

mhham reviewed (changes requested):
Mostly wording, and a few questions

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -508,24 +520,30 @@ def set_single_split_validation(self, split_ratio=0.8, stratified=True):
-            self._raw_settings["code"] = code
+            else:
+                self._raw_settings["code"] = code
         return self
```
Cons:
- Not very standard, including in our API client
- Kind of binding afterwards (removing it means breaking)

Pros:
- Allows chaining of setters
- In a notebook / REPL, will immediately show how it has affected the settings

mhham replied:
Concerning the pros:

* Chaining of setters is not so pythonic, from what I know, so I am not sure users would use this naturally ([python mailing list](https://mail.python.org/pipermail/python-dev/2003-October/038855.html) and [wiki article on fluent interface](https://en.wikipedia.org/wiki/Fluent_interface#Python))
* For the notebook / REPL, if we really need to show the resulting settings, we could print them instead ?

instanceofme replied:
We probably don't want to print in a setter, unwanted side effect. It is perfectly acceptable to me to say that people would need to be explicit:

```python
object.set_stuff(arg1, arg2)
object.set_other_stuff(arg3)
object
```

samronsin replied:
OK makes sense, removing the returned `selves` then !

samronsin replied:
Actually I'm having second thoughts... 
The initial rationale was to make the side effects of some setters more apparent to the users. Typically there are default kwargs that can modify parameters if already set to a non-default value, e.g. `set_random_search()` will automatically set the `seed` to `0` which may or may not be a surprise. Thoughts on this @aimeecoelho @du-phan ?

samronsin replied:
The other solution was to warn the user in the case I mentioned. Maybe that's the best solution after all ?

instanceofme replied:
IMO a warning in the setter's docstring is enough, as people this API is for fairly advanced users.

aimeecoelho replied:
I'd say as long as it is documented it's fine, in the docstring as Adrien says should work.

samronsin replied:
Deal then !

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -901,22 +923,38 @@ def __repr__(self):
+        :return: list of possible values for this hyperparameter
+        :rtype: list of str
+        """
+        return list(self._algo_settings[self.name]["values"].keys())
```
`accepted_values` is rather internal. `get_all_possible_values` was aligned with `get_all_possible_algorithm_names`, which we can't change for backward compatibility.

mhham replied:
OK for the name of the method. But the return value is not adapted to `SingleCategoryHyperparameterSettings` where `self._algo_settings[self.name]` is not a `dict`, but a single value

samronsin replied:
@mhham thanks for catching this, `get_all_possible_values` should return `self.accepted_values` !

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -923,6 +923,26 @@ def __init__(self, raw_settings, hyperparameter_search_params):
         self._hyperparameter_search_params = hyperparameter_search_params
         self._hyperparameters_registry = dict()
 
+    def __setattr__(self, key, value):
+        if key in {"_hyperparameters_registry", "_hyperparameter_search_params"} or key in {"enabled", "strategy"}:
+            # attributes and properties of the PredictionAlgorithmSettings object must be handled separately
+            super(PredictionAlgorithmSettings, self).__setattr__(key, value)
+        elif key in self._hyperparameters_registry or (key=="lambda_" and "lambda" in self._hyperparameters_registry):
+            if isinstance(value, HyperparameterSettings):
+                # call from a PredictionAlgorithmSettings child's __init__
+                super(PredictionAlgorithmSettings, self).__setattr__(key, value)
+            else:
+                # syntactic sugars
+                target = self._hyperparameters_registry[key]
+                if isinstance(target, (SingleValueHyperparameterSettings, SingleCategoryHyperparameterSettings)):
+                    target.set_value(value)
+                elif isinstance(target, CategoricalHyperparameterSettings):
+                    target.set_values(value)
+                else:
+                    raise Exception("Invalid type for assignment of a NumericalHyperparameterSettings object")
+        else:
+            raise Exception("Unknown hyperparameter: \"{}\"".format(key))
+
```
I don't think this works for the `__init__`, because the keys are added to the registry in the `__init__`, so they won't be there at first execution, and you will directly go to the `else` clause.

Can't we simply check if `hasattr(self, key)` to distinguish between the `__init__` and non `__init__` case. Then:
* If `not hasattr(self, key)` =>  We are in `__init__` (or `key` is a new attribute that we don't care about) => `super(PredictionAlgorithmSettings, self).__setattr__(key, value)`
* Otherwise => we are not in `__init__` for sure => check if key is in registry
    * if not in registry so this is not a hyperparam, do the standard thing => `super(PredictionAlgorithmSettings, self).__setattr__(key, value)` 
    * If in registry, this is a hyperparam=> check `target = self._hyperparameters_registry[key]`
        * If `target` is an instance of one of the three specific classes, use set value(s) method (and also would be nice to add type check in the `set_values` and `set_value` methods (it is already the case in `SingleValueHyperparameterSettings.set_value`) )
        * Otherwise raise an Exception, because we don't want the API users to manually modify the hyperparams
        
What do you think ?

samronsin replied:
Actually it does work because the call to `self._register_*_hyperparameter` (and thus the registration to `_hyperparameters_registry`) is executed _before_ the call to `__setattr__` in the children's `__init__`...
But the flow you're suggesting might be less ugly / more understandable and maintainable, so let me try and test it.

samronsin replied:
OK much better indeed IMO, thanks for the suggestion !

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +313,979 @@ def save(self):
+                target.set_value(value)
+            elif isinstance(target, CategoricalHyperparameterSettings):
+                target.set_values(value)
+            else:
+                raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
+        elif key == "lambda_":
+            raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
```
Why not set the `.values`? Not sure why `algo.cat = ["v1", "v2"]` would work but not `algo.num = [1, 2]`?

mhham replied:
I agree 👍 
Should we just set `.values` or `set_explicit_values` for `NumericalHyperparameterSettings`  ?

IMO should be `set_explicit_values` for the syntactic sugars, so that the user does not wonder why "nothing happens after setting values"

mhham replied:
> I disagree: having syntactic sugar for explicit values makes them look
privileged compared to ranges, and we would likely have it look the other
way around, especially to advanced users.

In that case it would be nice to have some syntactic sugar for both explicit and range values. Something like:
* If the value is a `list` => `set_explicit_values`
* if the value is a `Range` => `set_range`

But this would require to modify the `Range` class so that it is independent of the `NumericalHyperparameterSettings`. 

Not sure if it is worth it

instanceofme replied:
OK let's say that we can bench this for now, and not yet have direct assignment for multi-valued numerical HPs.

samronsin replied:
Created [CH58941](https://app.clubhouse.io/dataiku/story/58941/public-api-python-client-add-syntactic-sugar-for-numericalhyperparametersettings) to this purpose.

samronsin commented:
I disagree: having syntactic sugar for explicit values makes them look
privileged compared to ranges, and we would likely have it look the other
way around, especially to advanced users.

Le ven. 18 déc. 2020 à 12:40, Mehdi Hamoumi <notifications@github.com> a
écrit :

> *@mhham* commented on this pull request.
> ------------------------------
>
> In dataikuapi/dss/ml.py
> <https://github.com/dataiku/dataiku-api-client-python/pull/95#discussion_r545777377>
> :
>
> > +            else:
>
> +                raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
>
> +        elif key == "lambda_":
>
> +            raise Exception("Invalid assignment of a NumericalHyperparameterSettings object")
>
>
> I agree 👍
> Should we just set .values or set_explicit_values for
> NumericalHyperparameterSettings ?
>
> IMO should be set_explicit_values for the syntactic sugars, so that the
> user does not wonder why "nothing happens after setting values"
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dataiku/dataiku-api-client-python/pull/95#discussion_r545777377>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARC7VIRTZP2CQ4J2S2UDZ3SVM5SLANCNFSM4RQVDVEA>
> .
>

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -716,47 +798,39 @@ def __repr__(self):
+        :param values: values to enable, all other values will be disabled
+        :type values: list of str
+        """
+        all_possible_values = self.get_all_possible_values()
+        for category in values:
```
Add a check that values is indeed a `list`. If user wants to pass a single value (like me when testing 😬), the displayed error is complex to understand, as we iterate over the characters in that case.

```suggestion
        all_possible_values = self.get_all_possible_values()
        assert isinstance(values, list), \
            "Invalid input type {} for categorical hyperparameter {}: must be a list of strings".format(type(values), self.name)
        for category in values:
```

samronsin replied:
Good point, but we probably want to allow other iterables ? IIRC I had removed this check for that reason here (or maybe somewhere else).

nicolasservel replied:
To make things easy, we can just check that values is not some kind of string to allow people pass fancy iterables.

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +313,979 @@ def save(self):
+            "Invalid input type for hyperparameter \"{}\": ".format(self.name) + \
+            "range bounds must be numbers"
+
+    def _set_range(self, min=None, max=None, nb_values=None):
+        if min is None and max is None and nb_values is None:
+            warnings.warn("Numerical range for hyperparameter \"{}\" not modified".format(self.name))
+        else:
+            if min is not None:
+                self._check_number_input(min)
+                limit_min = self._algo_settings[self.name]["limit"].get("min")
+                if limit_min is not None:
+                    assert limit_min <= min, "Range min {} is below hyperparameter \"{}\" limit {}".format(min, self.name, limit_min)
+            if max is not None:
+                self._check_number_input(max)
+                limit_max = self._algo_settings[self.name]["limit"].get("max")
+                if limit_max is not None:
+                    assert max <= limit_max, "Range max {} is above hyperparameter \"{}\" limit {}".format(max, self.name, limit_max)
+            if nb_values is not None:
+                assert isinstance(nb_values, int) and nb_values >= 2, "Range number of values for hyperparameter \"{}\" must be an integer and >= 2".format(self.name)
+
+            if min is not None:
+                self._algo_settings[self.name]["range"]["min"] = min
+            if max is not None:
+                self._algo_settings[self.name]["range"]["max"] = max
+            if nb_values is not None:
+                self._algo_settings[self.name]["range"]["nbValues"] = nb_values
```
We do use variables with names `min` & `max`, that are overriding builtin functions. This does not break the code but I think it is not a good practice.

Besides, we do a lot of checks but we don't check that `min <= max`, if they are both defined

samronsin replied:
Using `min` and `max` naming offers the benefit of transparency (and arguably simplicity) to the user with the downside of a not-so-great implementation. The only real risk I can see with this is encouraging the user to use variables `min` and `max` themselves and then override the builtin functions, which could cause trouble down the line.
Using something like `maximum` and `minimum` as kwargs would not prevent this though... and would add a bit of complexity by introducing a mismatch between the json and the python object, so it's not clear to me which one is best for the user. Having the internal implementation not 100% clean (overriding `min` and `max` builtin functions in a scope where where is no chance they are being uses) is not a big deal IMO if it is better for the user, but the question is: is it ?

Regardless, we should probably check that `min <= max` indeed !

instanceofme replied:
Not a big deal indeed, I'd favor direct mapping to the JSON for the user's sake.

nicolasservel replied:
let's keep `min` & `max` then

nicolasservel reviewed (changes requested):
* In the docstring, there should be a blank line between the description and the params. Otherwise, it does not render well in the documentation.

* `disable_all_algorithms` does not work (it does not do anything for algos in algo_map). Shoud be replaced by something like:

```python
    def disable_all_algorithms(self):
        """Disables all algorithms"""

        for algorithm_key in self.__class__.algorithm_remap.keys():
            algorithm_meta = self.__class__.algorithm_remap[algorithm_key]
            if isinstance(algorithm_meta, PredictionAlgorithmMeta):
                key = algorithm_meta.algorithm_name
            else:
                key = algorithm_meta
            if key in self.mltask_settings["modeling"]:
                self.mltask_settings["modeling"][key]["enabled"] = False

        for custom_mllib in self.mltask_settings["modeling"]["custom_mllib"]:
            custom_mllib["enabled"] = False
        for custom_python in self.mltask_settings["modeling"]["custom_python"]:
            custom_python["enabled"] = False
        for plugin in self.mltask_settings["modeling"]["plugin_python"].values():
            plugin["enabled"] = False
```

Note it would be more elegant to have an `AlgorithmMeta` that we had at some point rather than a `PredictionAlgorithmMeta` that forces to distinguish cases in the parent class.

instanceofme commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +313,979 @@ def save(self):
+        self.booster = self._register_categorical_hyperparameter("booster")
+        self.objective = self._register_categorical_hyperparameter("objective")
+        self.n_estimators = self._register_single_value_hyperparameter("n_estimators", accepted_types=[int])
+        self.nthread = self._register_single_value_hyperparameter("nthread", accepted_types=[int])
+        self.scale_pos_weight = self._register_single_value_hyperparameter("scale_pos_weight", accepted_types=[int, float])
+        self.base_score = self._register_single_value_hyperparameter("base_score", accepted_types=[int, float])
+        self.impute_missing = self._register_single_value_hyperparameter("impute_missing", accepted_types=[bool])
+        self.missing = self._register_single_value_hyperparameter("missing", accepted_types=[int, float])
+        self.cpu_tree_method = self._register_single_category_hyperparameter("cpu_tree_method", accepted_values=["auto", "exact", "approx", "hist"])
+        self.gpu_tree_method = self._register_single_category_hyperparameter("gpu_tree_method", accepted_values=["gpu_exact", "gpu_hist"])
+        self.enable_cuda = self._register_single_value_hyperparameter("enable_cuda", accepted_types=[bool])
+        self.seed = self._register_single_value_hyperparameter("seed", accepted_types=[int])
```
For `nthread` is not a hyperparameter and will never be searchable (as it has theoretically no influence on results), so unlike other single-valued hyperparameters, it should probably be a simple gettable + settable property, meaning getting it should return the direct value instead of a "single-valued hyperparameter" object.

It is probably the same for:
- `n_jobs` on algorithms settings
- `seed` on algorithms settings? Could that ever be searchable?

mhham replied:
I agree, but with the current hyperparams registration mechanism, these parameters would not be printed in the `__repr__(html__)` methods if they are defined as separate properties

instanceofme replied:
Nice catch, that would need to be handled manually as well.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self._raw_settings = raw_settings
+
+    def _key_repr(self, key):
+        if is_basestring(self._raw_settings[key]):
```
The idea here is to print quotes around strings only. Some of the parameters' values are boolean (e.g. `"stratified"`) or numbers (e.g. `"nFolds"`)

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +313,979 @@ def save(self):
+        else:
+            return "    \"{}\"={}\n".format(key, self._raw_settings[key])
+
+    def _repr_html_(self):
```
I assume you meant in the `PredictionAlgorithmSettings` class.

instanceofme replied:
(and clustering too then)

samronsin replied:
Actually we don't pretty print anything for clustering algorithms, so not sure what you mean @instanceofme.

instanceofme replied:
Ok, disregard my question then 😅

samronsin commented:
@mhham @nicolasservel @instanceofme I think I addressed all your remarks, let me know if you find anything else to improve, otherwise let's wrap this up. Thanks a lot for your comments, bug reports and suggestions !

instanceofme commented:
Looking good wrt what we discussed. Would you mind updating the example notebook?

samronsin commented:
New notebook in [CH53507](https://app.clubhouse.io/dataiku/story/53507/improve-hyper-parameter-search-experience-in-python-client-of-ml-api-for-each-algorithm).

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -801,9 +816,12 @@ def _pretty_repr(self):
+        :param values: list of values to enable, all other values will be disabled
         :type values: list of str
         """
+        assert isinstance(values, list) and not isinstance(values, string_types), \
```
Isn't `isinstance(values, list)` enough ?

samronsin replied:
Hum yes, indeed... I once considered testing more largely against iterables and thus explicitly exclude strings but after all it's so easy to make any iterable a `list` that I decided to stick to `list`.

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -915,53 +933,78 @@ class PredictionAlgorithmSettings(dict):
+                else:
+                    # simple parameter
+                    assert isinstance(value, type(target)), "Invalid type {} for parameter {}: expected {}".format(type(value), attr_name, type(target))
+                    super(PredictionAlgorithmSettings, self).__setattr__(attr_name, value)
+                    self[attr_name] = value
```
Shouldn't the first line suffice ?

samronsin replied:
Actually no, e.g. for attribute `epsilon`:
- `super(PredictionAlgorithmSettings, self).__setattr__(attr_name, value)` handles the attribute `epsilon` of the `PredictionAlgorithmSettings` object
- `self[attr_name] = value` handles the dict value at key `"epsilon"` which is different from the attribute

mhham replied:
Yes, you're right, I got confused my bad

instanceofme replied:
Maybe it warrants a comment to make sure nobody accidentally removes that second line?

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+            self._attr_to_json_remapping[attr_name] = json_key
+
+    def _register_numerical_hyperparameter(self, json_key, attr_name=None):
+        self._maybe_register_attr_json_mismatch(json_key, attr_name)
```
Might be a tiny bit more readable to rename method to `_register_attr_json_mismatch` and 
```suggestion
        if attr_name is not None:
            self._register_attr_json_mismatch(json_key, attr_name)
```

samronsin replied:
Why not full inline instead of creating a method then ?
```suggestion
        if attr_name is not None:
            self._attr_to_json_remapping[attr_name] = json_key
```

mhham commented:
Looking good !

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+        return self._hyperparameters_registry[json_key]
+
+    def _repr_html_(self):
+        res = "<pre>" + self.__class__.__name__ + "\n"
```
Add parentheses and indentation after class name ?

mhham commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+    def _repr_html_(self):
+
+        res = "<pre>"
+        res += self.__class__.__name__ + "(\n"
```
Add indentation after class name ?

nicolasservel commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+            "Invalid input type for hyperparameter \"{}\": ".format(self.name) + \
+            "range bounds must be numbers"
+
+    def _set_range(self, min=None, max=None, nb_values=None):
```
did not add check on `min <= max`

nicolasservel reviewed (changes requested):
last remarks

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+        else:
+            if attr_name in self._attr_to_json_remapping:
+                # attribute name and json key mismatch (e.g. "lambda", "alphaMode")
+                attr_name = self._attr_to_json_remapping[attr_name]
```
I was thinking the same thing and updated accordingly.

samronsin commented on dataikuapi/dss/ml.py:
```diff
@@ -315,43 +319,1016 @@ def save(self):
+        self.C = self._register_numerical_hyperparameter("C")
+        self.penalty = self._register_categorical_hyperparameter("penalty")
+        self.multi_class = self._register_single_category_hyperparameter("multi_class", accepted_values=["multinomial", "ovr"])
+        self.n_jobs = self._register_single_value_hyperparameter("n_jobs", accepted_types=[int])
```
Indeed !

instanceofme reviewed (approved):
LGTM. One last remark from Clément:

In str representation, we could rename `gridMode` to `definition_mode` (used as setter/getter) and leave `gridMode` in the Details (if one wants to see the actual JSON/dict).

samronsin commented:
> In str representation, we could rename gridMode to definition_mode (used as setter/getter) and leave gridMode in the Details (if one wants to see the actual JSON/dict).

In the `_repr_html` of `NumericalHyperparameterSettings` then ?

instanceofme commented:
> > In str representation, we could rename gridMode to definition_mode (used as setter/getter) and leave gridMode in the Details (if one wants to see the actual JSON/dict).
> 
> In the `_repr_html` of `NumericalHyperparameterSettings` then ?

Yes (only for the "filtered" view, not the Details)
